File: elixir/saddle-points/saddle_points.exs
defmodule Matrix do

  @doc """
  Parses a string representation of a matrix
  to a list of rows
  """
  @spec rows(String.t()) :: [[integer]]
  def rows(str) do
    str
    |> String.split("\n")
    |> Enum.map(&parse_row/1)
  end

  defp parse_row(str) do
    str
    |> String.split
    |> Enum.map(&String.to_integer/1)
  end

  @doc """
  Parses a string representation of a matrix
  to a list of columns
  """
  @spec columns(String.t()) :: [[integer]]
  def columns(str) do
    str
    |> rows
    |> List.zip
    |> Enum.map(&Tuple.to_list/1)
  end

  @doc """
  Calculates all the saddle points from a string
  representation of a matrix
  """
  @spec saddle_points(String.t()) :: [{integer, integer}]
  def saddle_points(str) do
    cols = smallest_in_columns(str)
    rows = largest_in_rows(str)

    Enum.filter(rows, fn(row) -> row in cols end)
  end

  defp largest_in_rows(str) do
    str
    |> rows
    |> Enum.with_index
    |> Enum.reduce([], fn({row,idx}, acc) -> acc ++ largest_in_row(row, idx) end)
  end

  defp largest_in_row(row, row_num) do
    largest = row |> Enum.max
    row
    |> Enum.with_index
    |> Enum.filter_map(fn({val,_}) -> val == largest end, fn({_,col}) -> {row_num,col} end)
  end

  defp smallest_in_columns(str) do
    str
    |> columns
    |> Enum.with_index
    |> Enum.reduce([], fn({col,idx}, acc) -> acc ++ smallest_in_column(col,idx) end)
  end

  defp smallest_in_column(col, col_num) do
    smallest = col |> Enum.min
    col
    |> Enum.with_index
    |> Enum.filter_map(fn({val,_}) -> val == smallest end, fn({_,row}) -> {row,col_num} end)
  end

end

File: go/saddle-points/saddle_points.go
package matrix

import (
	"errors"
	"strconv"
	"strings"
)

type Pair struct {
	row int
	col int
}

type Matrix struct {
	data [][]int
}

func (m *Matrix) Saddle() []Pair {
	points := []Pair{}

	for r, row := range m.data {
		for c, val := range row {
			if m.smallestInColumn(c, val) && m.largestInRow(r, val) {
				points = append(points, Pair{r, c})
			}
		}
	}

	return points
}

func (m *Matrix) smallestInColumn(col int, val int) bool {
	for _, row := range m.data {
		if row[col] < val {
			return false
		}
	}

	return true
}

func (m *Matrix) largestInRow(row int, val int) bool {
	for _, v := range m.data[row] {
		if v > val {
			return false
		}
	}

	return true
}

func New(input string) (*Matrix, error) {
	lines := strings.Split(input, "\n")
	data := make([][]int, 0, len(lines))

	for _, line := range lines {
		tokens := strings.Fields(line)
		nums := make([]int, 0, len(tokens))

		for _, num := range tokens {
			if num == "" {
				continue
			}

			val, err := strconv.Atoi(num)
			if err != nil {
				return nil, errors.New("Invalid matrix")
			}

			nums = append(nums, val)
		}

		data = append(data, nums)
	}

	// make sure all arrays are the same length
	cols := len(data[0])
	for i := 1; i < len(data); i++ {
		if len(data[i]) != cols {
			return nil, errors.New("Invalid matrix")
		}
	}

	return &Matrix{data: data}, nil
}
